### ISSUE TITLE
🚀 Feature Request: Enable code-splitting in the deploy esbuild config

### ISSUE THREAD

**Issue by dmaretskyi** [cloudflare/workers-sdk#7344]
### Describe the solution

**Description**  
Currently, when deploying a Worker using Wrangler, all code is bundled into a single file. For larger projects, this can result in significant startup delays due to the inclusion of code that may not be needed for every execution path. Dynamic imports could help mitigate this by allowing unused code to be split into separate chunks and loaded only when required. However, Wrangler does not currently support this behavior during deployment. 

**Use Case**  
In scenarios where the Worker bundle grows large, startup times can negatively impact performance. Code splitting would allow developers to manage their Worker bundles more efficiently by:  
- Reducing initial payload size.  
- Improving execution performance by loading code only when necessary.  

**Challenges**  
Although the bundling documentation mentions the `additionalModules` configuration as a potential workaround, this approach feels more complex than expected for a common use case like code splitting. Having built-in support for code splitting during deployment would simplify workflows significantly.  

**Proposed Solution**  
Enable support for ESBuild's code-splitting capabilities in Wrangler. This could involve updating the deploy bundle configuration to:  
1. Automatically detect dynamic imports.  
2. Generate separate chunks for dynamically imported modules.  
3. Ensure the runtime environment supports fetching and executing these chunks.  

**Benefits**  
- Improved Worker startup times.  
- Smaller initial bundle sizes.  
- Easier optimization for projects with modular architecture.

**Questions for Maintainers**  
- Are there any technical or runtime constraints preventing this feature from being implemented?  
- Could this be added as an optional configuration in `wrangler.toml` for users who want to opt-in to code splitting?

**Comment by vicb** [comment]
> **Benefits**
> 
> * Improved Worker startup times.
> * Smaller initial bundle sizes.
> * Easier optimization for projects with modular architecture.

Currently all the code (main bundle + async loaded external modules) are parsed at startup. Code splitting should not show major perf differences.

> Improved Worker startup times

As explained earlier, this should not make a major difference.

It is possible to tests by building the code with external modules and then uploading it using `--no-bundle`.
Please report if you see an improvement.

> Smaller initial bundle sizes

Not sure what this brings, can you expand ?

> Easier optimization for projects with modular architecture

Can you expand on what you mean here ?

**Comment by ItsWendell** [comment]
Fair enough, I didn't realize that in https://github.com/cloudflare/workerd/issues/2372 someone ran a benchmark for code / bundle split workers / pages.

It sounds like it would still be beneficial to have perf optimizations here in workerd, but until that time this PR is less relevant: https://github.com/cloudflare/workers-sdk/pull/9240

**Comment by penalosa** [comment]
I'm going to close this for now. It's still blocked on underlying workerd improvements, and going forward we'd generally recommend [`@cloudflare/vite-plugin`](https://developers.cloudflare.com/workers/vite-plugin/) as a way to deeply customise your build setup.
